File: app/src/main/java/chat/rocket/android/fragment/chatroom/RoomContract.java
package chat.rocket.android.fragment.chatroom;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import java.util.List;

import chat.rocket.android.shared.BaseContract;
import chat.rocket.android.widget.AbsoluteUrl;
import chat.rocket.core.models.Message;
import chat.rocket.core.models.Room;
import chat.rocket.core.models.User;

public interface RoomContract {

  interface View extends BaseContract.View {

    void setupWith(RocketChatAbsoluteUrl rocketChatAbsoluteUrl);

    void render(Room room);

    void showUserStatus(User user);

    void updateHistoryState(boolean hasNext, boolean isLoaded);

    void onMessageSendSuccessfully();

    void disableMessageInput();

    void enableMessageInput();

    void showUnreadCount(int count);

    void showMessages(List<Message> messages);

    void showMessageSendFailure(Message message);

    void showMessageDeleteFailure(Message message);

    void autoloadImages();

    void manualLoadImages();

    void onReply(AbsoluteUrl absoluteUrl, String markdown, Message message);

    void onCopy(String message);

    void showMessageActions(Message message);
  }

  interface Presenter extends BaseContract.Presenter<View> {

    void loadMessages();

    void loadMoreMessages();

    void onMessageSelected(@Nullable Message message);

    void onMessageTap(@Nullable Message message);

    void sendMessage(String messageText);

    void resendMessage(@NonNull Message message);

    void updateMessage(@NonNull Message message, String content);

    void deleteMessage(@NonNull Message message);

    void onUnreadCount();

    void onMarkAsRead();

    void refreshRoom();

    void replyMessage(@NonNull Message message, boolean justQuote);

    void acceptMessageDeleteFailure(Message message);
  }
}

File: app/src/main/java/chat/rocket/android/fragment/chatroom/RoomContract.kt
package chat.rocket.android.fragment.chatroom

import chat.rocket.android.shared.BaseContract
import chat.rocket.android.widget.AbsoluteUrl
import chat.rocket.core.models.Message
import chat.rocket.core.models.Room
import chat.rocket.core.models.User

interface RoomContract {

    interface View : BaseContract.View {

        fun setupWith(rocketChatAbsoluteUrl: RocketChatAbsoluteUrl)

        fun render(room: Room)

        fun showUserStatus(user: User)

        fun updateHistoryState(hasNext: Boolean, isLoaded: Boolean)

        fun onMessageSendSuccessfully()

        fun disableMessageInput()

        fun enableMessageInput()

        fun showUnreadCount(count: Int)

        fun showMessages(messages: List<Message>)

        fun showMessageSendFailure(message: Message)

        fun showMessageDeleteFailure(message: Message)

        fun autoloadImages()

        fun manualLoadImages()

        fun onReply(absoluteUrl: AbsoluteUrl, markdown: String, message: Message)

        fun onCopy(message: String)

        fun showMessageActions(message: Message)
    }

    interface Presenter : BaseContract.Presenter<View> {

        fun loadMessages()

        fun loadMoreMessages()

        fun onMessageSelected(message: Message?)

        fun onMessageTap(message: Message?)

        fun sendMessage(messageText: String)

        fun resendMessage(message: Message)

        fun updateMessage(message: Message, content: String)

        fun deleteMessage(message: Message)

        fun onUnreadCount()

        fun onMarkAsRead()

        fun refreshRoom()

        fun replyMessage(message: Message, justQuote: Boolean)

        fun acceptMessageDeleteFailure(message: Message)

        fun loadMissedMessages()
    }
}
